File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/controller/IndexController.java
package com.windea.demo.cloudcollect.core.controller;

import com.windea.demo.cloudcollect.core.domain.entity.Collect;
import com.windea.demo.cloudcollect.core.domain.entity.User;
import com.windea.demo.cloudcollect.core.domain.model.JwtUserDetails;
import com.windea.demo.cloudcollect.core.domain.view.EmailRegisterView;
import com.windea.demo.cloudcollect.core.domain.view.UsernamePasswordLoginView;
import com.windea.demo.cloudcollect.core.service.CollectService;
import com.windea.demo.cloudcollect.core.service.UserService;
import io.swagger.annotations.*;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.security.Principal;

/**
 * TODO 首页的控制器。包含登录、注册等操作。
 */
@Api("首页")
@RestController
@RequestMapping("/")
@CrossOrigin
public class IndexController {
	private final CollectService collectService;
	private final UserService userService;

	public IndexController(CollectService collectService, UserService userService) {
		this.collectService = collectService;
		this.userService = userService;
	}


	@ApiOperation("通过用户名&密码登录用户。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "view", value = "用户名&密码登录视图", required = true)
	})
	@PostMapping({"/login", "/loginByUsernameAndPassword"})
	public User loginByUsernameAndPassword(@RequestBody @Valid UsernamePasswordLoginView view,
		BindingResult bindingResult) {
		return userService.loginByUsernameAndPassword(view);
	}

	@ApiOperation("通过邮箱注册用户。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "view", value = "邮箱注册视图", required = true)
	})
	@PostMapping({"/register", "/registerByEmail"})
	public User registerByEmail(@RequestBody @Valid EmailRegisterView view, BindingResult bindingResult) {
		return userService.registerByEmail(view);
	}

	@ApiOperation("激活用户。")
	@PutMapping("/activate")
	public User activate(Principal principal) {
		var user = ((JwtUserDetails) principal).getDelegateUser();
		return userService.activate(user);
	}

	@ApiOperation("重置用户密码。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "newPassword", value = "新的密码", required = true)
	})
	@PutMapping("/resetPassword")
	public User resetPassword(@RequestParam String newPassword, Principal principal) {
		var user = ((JwtUserDetails) principal).getDelegateUser();
		return userService.resetPassword(user, newPassword);
	}

	@ApiOperation("随便看看任一收藏。")
	@GetMapping("/lookAroundCollect")
	public Collect lookAroundCollect() {
		return collectService.getByRandom();
	}

	@ApiOperation("随便看看任一用户信息")
	@GetMapping("/lookAroundUser")
	public User lookAroundUser() {
		return userService.getByRandom();
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/controller/IndexController.kt
@file:Suppress("UNUSED_PARAMETER")

package com.windea.demo.cloudcollect.core.controller

import com.windea.demo.cloudcollect.core.domain.entity.*
import com.windea.demo.cloudcollect.core.domain.request.*
import com.windea.demo.cloudcollect.core.domain.response.*
import com.windea.demo.cloudcollect.core.service.*
import com.windea.demo.cloudcollect.core.validation.group.*
import io.swagger.annotations.*
import org.springframework.validation.*
import org.springframework.validation.annotation.*
import org.springframework.web.bind.annotation.*

//包含登录注册功能，以及随便看看功能
@Api("首页")
@RestController
@RequestMapping("/")
@CrossOrigin
class IndexController(
	private val collectService: CollectService,
	private val userService: UserService
) {
	@ApiOperation("登录用户。")
	@PostMapping("/login")
	fun login(@RequestBody @Validated form: LoginForm, bindingResult: BindingResult): UserDetailsVo {
		return userService.login(form)
	}
	
	@ApiOperation("注册用户。")
	@PostMapping("/register")
	fun register(@RequestBody @Validated(Create::class) user: User, bindingResult: BindingResult) {
		userService.register(user)
	}
	
	@ApiOperation("激活用户。")
	@PostMapping("/activate")
	fun activate(@RequestParam username: String, @RequestParam activateCode: String) {
		userService.activate(username, activateCode)
	}
	
	@ApiOperation("忘记用户密码。")
	@PostMapping("/forgotPassword")
	fun forgotPassword(@RequestParam username: String) {
		userService.forgotPassword(username)
	}
	
	@ApiOperation("重置用户密码。")
	@PostMapping("/resetPassword")
	fun resetPassword(@RequestBody @Validated form: ResetPasswordForm, bindingResult: BindingResult,
		@RequestParam resetPasswordCode: String) {
		userService.resetPassword(form, resetPasswordCode)
	}
	
	
	@ApiOperation("随便看看任一收藏。")
	@GetMapping("/lookAroundCollect")
	fun lookAroundCollect(): Collect {
		return collectService.findByRandom()
	}
	
	@ApiOperation("随便看看任一用户。")
	@GetMapping("/lookAroundUser")
	fun lookAroundUser(): User {
		return userService.findByRandom()
	}
}
